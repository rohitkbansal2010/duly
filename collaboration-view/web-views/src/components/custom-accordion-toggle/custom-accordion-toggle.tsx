import { useContext } from 'react';
import { AccordionContext, useAccordionButton } from 'react-bootstrap';

type CustomAccordionTogglePropsType = {
  eventKey: string;
  callback: (isCurrentEventKey: boolean) => void;
  children: (
    decoratedOnClick: ReturnType<typeof useAccordionButton>,
    isCurrentEventKey: boolean
  ) => JSX.Element;
};

export const CustomAccordionToggle = ({
  eventKey,
  callback,
  children,
}: CustomAccordionTogglePropsType) => {
  const { activeEventKey } = useContext(AccordionContext);

  const isCurrentEventKey = activeEventKey === eventKey;

  const decoratedOnClick = useAccordionButton(
    eventKey,
    () =>
      callback(isCurrentEventKey)
  );

  return children(decoratedOnClick, isCurrentEventKey);
};
