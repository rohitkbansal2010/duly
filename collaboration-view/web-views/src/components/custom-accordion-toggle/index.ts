export { CustomAccordionToggle } from './custom-accordion-toggle';
