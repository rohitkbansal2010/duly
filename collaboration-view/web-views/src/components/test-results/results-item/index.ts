export { ResultsItem } from './results-item';
