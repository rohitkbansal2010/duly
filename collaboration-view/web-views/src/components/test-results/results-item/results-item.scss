@import "~@styles/typography";
@import "~@styles/colors";
@import "~../results-components/results-components.scss";

.results {
  &Components {
    margin-top: 1.25rem;

    &Item {
      display: flex;
      background: $light-gray;
      border: 0.0625rem solid $gray;
      border-top: none;

      &Accordion {
        &:last-child {
          .resultsComponentsItem {
            border-radius: 0 0 0.5rem 0.5rem;
          }
        }
      }

      &Abnormal {
        background: rgba($dupage-orange, 0.05);
      }

      &Name {
        @include inter-text-default-bold;
        @include name;
        @include component-item;
        @include border-right;
        font-size: var(--font-size-16px-4px);
      }

      &Value {
        @include inter-text-default;
        @include value;
        @include component-item;
        @include border-right;
        font-size: var(--font-size-16px-4px);

        img {
          position: relative;
          top: -0.625rem;
          width: 1.25rem;
        }
      }

      &Range {
        @include inter-text-default;
        @include range;
        @include component-item;
        @include border-right;
        font-size: var(--font-size-16px-4px);
      }

      &Additional {
        @include additional;
        @include component-item;

        display: flex;
        justify-content: space-between;

        &Button {
          @include inter-text-default-bold;
          display: flex;
          align-items: center;
          font-size: var(--font-size-14px-4px);
          line-height: 1.375rem;
          text-transform: capitalize;
          background: none;
          border: none;

          img {
            width: 1.125rem;
            margin-right: 0.5rem;
          }
        }
      }

      &Expand {
        padding: 1.25rem;
        border: 0.0625rem solid $gray;
        border-top: none;

        &Title {
          @include inter-text-default-bold;
          font-size: var(--font-size-16px-4px);
          line-height: var(--font-size-20px-4px);

          text-transform: capitalize;
        }

        &Text {
          @include inter-text-default;
          font-size: var(--font-size-16px-4px);
          white-space: pre-line;
        }
      }
    }
  }
}
