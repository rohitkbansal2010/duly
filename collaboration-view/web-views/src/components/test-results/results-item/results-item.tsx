import { Accordion } from 'react-bootstrap';

import { CustomAccordionToggle } from '@components/custom-accordion-toggle';
import { LevelsChart } from '@components/levels-chart';
import { TEST_RESULTS_CHART_HEIGHT, NO_SHOWING_TEXT } from '@constants';
import { TestResultsAccordions } from '@enums';
import {
  noteResultDarkBlueIcon,
  pastResultLightBlueIcon,
  speechBubbleOrangeIcon
} from '@icons';
import { CHART_MOCK_DATA } from '@mock-data';
import { TestResultMeasurement, ReferenceRange } from '@types';
import {
  formattedInterpretations,
  formattedNotes,
  getMeasurementUnit,
  getMeasurementValue,
  getReferenceRange
} from '@utils';

import styles from './results-item.scss';

type ResultsComponentProps = {
  componentName: string;
  isAbnormalResult: boolean;
  measurement?: TestResultMeasurement;
  referenceRange?: ReferenceRange;
  valueText?: string;
  notes?: string[];
  interpretations?: string[];
  eventKey: string;
  activeKey: string;
  onClickNotes: () => void;
  onClickPastResults: () => void;
}

export const ResultsItem = ({
  componentName,
  isAbnormalResult,
  measurement,
  referenceRange,
  valueText = '',
  notes = [],
  interpretations = [],
  activeKey,
  eventKey,
  onClickNotes,
  onClickPastResults,
}: ResultsComponentProps) => {
  const resultComponentsItemClasses = []
    .concat(styles.resultsComponentsItem)
    .concat(isAbnormalResult ? styles.resultsComponentsItemAbnormal : '')
    .join(' ');

  const isNotesShown = Boolean(notes.length || interpretations.length);

  return (
    <Accordion
      activeKey={activeKey}
      className={styles.resultsComponentsItemAccordion}
    >
      <div className={resultComponentsItemClasses}>
        <div className={styles.resultsComponentsItemName}>
          {componentName}
        </div>
        <div className={styles.resultsComponentsItemValue}>
          {(!valueText.toLowerCase().includes(NO_SHOWING_TEXT) && valueText) ||
            (measurement && `${getMeasurementValue(measurement)} ${getMeasurementUnit(measurement)}`)}
          {isAbnormalResult && <img src={speechBubbleOrangeIcon} alt="abnormal result" />}
        </div>
        <div className={styles.resultsComponentsItemRange}>
          {getReferenceRange(referenceRange)}
        </div>
        <div className={styles.resultsComponentsItemAdditional}>
          <CustomAccordionToggle
            eventKey={`${eventKey}-${TestResultsAccordions.PAST_RESULTS}`}
            callback={onClickPastResults}
          >
            {decoratedOnClick =>
                (
                  <button
                    className={styles.resultsComponentsItemAdditionalButton}
                    onClick={decoratedOnClick}
                  >
                    <img src={pastResultLightBlueIcon} alt="past results" />
                    <span>past results</span>
                  </button>
                )}
          </CustomAccordionToggle>
          {isNotesShown && (
            <CustomAccordionToggle
              eventKey={`${eventKey}-${TestResultsAccordions.NOTES}`}
              callback={onClickNotes}
            >
              {decoratedOnClick =>
                (
                  <button
                    className={styles.resultsComponentsItemAdditionalButton}
                    onClick={decoratedOnClick}
                  >
                    <img src={noteResultDarkBlueIcon} alt="notes" />
                    <span>notes</span>
                  </button>
                )}
            </CustomAccordionToggle>
          )}
        </div>
      </div>
      <Accordion.Collapse
        eventKey={`${eventKey}-${TestResultsAccordions.NOTES}`}
      >
        <div className={styles.resultsComponentsItemExpand}>
          <span className={styles.resultsComponentsItemExpandTitle}>notes</span>
          <div className={styles.resultsComponentsItemExpandText}>
            {formattedInterpretations(interpretations)}
            {formattedNotes(notes)}
          </div>
        </div>
      </Accordion.Collapse>
      <Accordion.Collapse
        eventKey={`${eventKey}-${TestResultsAccordions.PAST_RESULTS}`}
      >
        <div className={styles.resultsComponentsItemExpand}>
          <div className={styles.resultsComponentsItemExpandChart}>
            <LevelsChart
              chartId="past result chart"
              chartData={CHART_MOCK_DATA}
              height={TEST_RESULTS_CHART_HEIGHT}
            />
          </div>
        </div>
      </Accordion.Collapse>
    </Accordion>
  );
};
