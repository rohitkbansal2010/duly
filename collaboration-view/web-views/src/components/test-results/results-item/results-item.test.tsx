import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { CustomAccordionToggle } from '@components/custom-accordion-toggle';
import { speechBubbleOrangeIcon } from '@icons';
import {
  getMeasurementValue,
  getMeasurementUnit,
  getReferenceRange,
  formattedNotes
} from '@utils';

import { ResultsItem } from './results-item';

describe('ResultsItem', () => {
  let component: ReturnType<typeof shallow>;

  const measurement = {
    value: 150,
    unit: 'mg/dL', 
  };

  const onClickNotes = jest.fn();
    
  const mockProps = {
    componentName: 'Cholesterol, Total',
    isAbnormalResult: true,
    measurement,
    referenceRange: { text: '<200' },
    notes: [
      'Desirable  <200 mg/dL\\r\\nBorderline  200-239 mg/dL\\r\\nHigh      >=240 mg/dL\\r\\n\\r\\n',
    ],
    activeKey: '06cd80cf-7230-4e53-bdd1-6956402b2ef8',
    eventKey: '06cd80cf-7230-4e53-bdd1-6956402b2ef8',
    onClickNotes,
  };

  const mockEmptyProps = {
    componentName: '',
    isAbnormalResult: false,
    activeKey: '06cd80cf-7230-4e53-bdd1-6956402b2ef8',
    eventKey: '06cd80cf-7230-4e53-bdd1-6956402b2ef8',
    onClickNotes,
  };

  beforeEach(() => {
    component = shallow(<ResultsItem {...mockProps} />);
  });

  it('component should match snapshot', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('component should contain right data', () => {
    const value =
      `${getMeasurementValue(mockProps.measurement)} ${getMeasurementUnit(mockProps.measurement)}`;
    const referenceRange = getReferenceRange(mockProps.referenceRange);
    const notes = formattedNotes(mockProps.notes);

    expect(component.text().includes(value)).toBeTruthy();
    expect(component.text().includes(mockProps.componentName)).toBeTruthy();
    expect(component.text().includes(referenceRange)).toBeTruthy();
    expect(component.text().includes(notes)).toBeTruthy();
  });

  it('component should contain notes and past results toggles', () => {
    expect(component.find(CustomAccordionToggle)).toHaveLength(2);
  });

  it('component should contain abnormal icon', () => {
    expect(component.findWhere(n =>
      n.type() === 'img' && n.props().src === speechBubbleOrangeIcon)).toHaveLength(1);
  });

  it('should work with empty data', () => {
    component = shallow(<ResultsItem {...mockEmptyProps}/>);

    expect(component.text().includes('undefined')).toBeFalsy();
    expect(toJson(component)).toMatchSnapshot();
  });
});
