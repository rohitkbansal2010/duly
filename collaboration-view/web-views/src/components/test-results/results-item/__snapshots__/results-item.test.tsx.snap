// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultsItem component should match snapshot 1`] = `
<Accordion
  activeKey="06cd80cf-7230-4e53-bdd1-6956402b2ef8"
  className="resultsComponentsItemAccordion"
>
  <div
    className="resultsComponentsItem resultsComponentsItemAbnormal"
  >
    <div
      className="resultsComponentsItemName"
    >
      Cholesterol, Total
    </div>
    <div
      className="resultsComponentsItemValue"
    >
      150 mg/dL
      <img
        alt="abnormal result"
        src="speech-bubble-orange.svg"
      />
    </div>
    <div
      className="resultsComponentsItemRange"
    >
      &lt;200
    </div>
    <div
      className="resultsComponentsItemAdditional"
    >
      <CustomAccordionToggle
        eventKey="06cd80cf-7230-4e53-bdd1-6956402b2ef8-PAST_RESULTS"
      >
        <Component />
      </CustomAccordionToggle>
      <CustomAccordionToggle
        callback={[MockFunction]}
        eventKey="06cd80cf-7230-4e53-bdd1-6956402b2ef8-NOTES"
      >
        <Component />
      </CustomAccordionToggle>
    </div>
  </div>
  <AccordionCollapse
    eventKey="06cd80cf-7230-4e53-bdd1-6956402b2ef8-NOTES"
  >
    <div
      className="resultsComponentsItemExpand"
    >
      <span
        className="resultsComponentsItemExpandTitle"
      >
        notes
      </span>
      <div
        className="resultsComponentsItemExpandText"
      >
        Desirable  &lt;200 mg/dL
Borderline  200-239 mg/dL
High      &gt;=240 mg/dL
      </div>
    </div>
  </AccordionCollapse>
  <AccordionCollapse
    eventKey="06cd80cf-7230-4e53-bdd1-6956402b2ef8-PAST_RESULTS"
  >
    <div
      className="resultsComponentsItemExpand"
    >
      <div
        className="resultsComponentsItemExpandChart"
      >
        <LevelsChart
          chartData={
            Object {
              "chartOptions": Object {
                "chartScales": Object {
                  "yAxis": Object {
                    "max": 110,
                    "min": 70,
                    "stepSize": 10,
                  },
                },
              },
              "datasets": Array [
                Object {
                  "data": Object {
                    "dimension": "mg/dl",
                    "values": Array [
                      Object {
                        "x": "2021-12-15T22:44:00",
                        "y": 90.7,
                      },
                      Object {
                        "x": "2021-11-15T22:44:00",
                        "y": 95.7,
                      },
                      Object {
                        "x": "2021-10-15T22:44:00",
                        "y": 97.7,
                      },
                      Object {
                        "x": "2021-09-15T22:44:00",
                        "y": 100.7,
                      },
                      Object {
                        "x": "2021-08-15T22:44:00",
                        "y": 105.7,
                      },
                      Object {
                        "x": "2021-07-15T22:44:00",
                        "y": 95.7,
                      },
                      Object {
                        "x": "2021-06-15T22:44:00",
                        "y": 108.7,
                      },
                      Object {
                        "x": "2021-05-15T22:44:00",
                        "y": 107.7,
                      },
                      Object {
                        "x": "2021-04-15T22:44:00",
                        "y": 80.7,
                      },
                      Object {
                        "x": "2021-03-15T22:44:00",
                        "y": 85.7,
                      },
                      Object {
                        "x": "2021-02-15T22:44:00",
                        "y": 90.7,
                      },
                      Object {
                        "x": "2021-01-15T22:44:00",
                        "y": 100.7,
                      },
                      Object {
                        "x": "2022-01-15T22:44:00",
                        "y": 100.7,
                      },
                      Object {
                        "x": "2022-02-15T22:44:00",
                        "y": 100.7,
                      },
                    ],
                  },
                  "label": "Past Results",
                },
              ],
            }
          }
          chartId="past result chart"
          height={310}
        />
      </div>
    </div>
  </AccordionCollapse>
</Accordion>
`;

exports[`ResultsItem should work with empty data 1`] = `
<Accordion
  activeKey="06cd80cf-7230-4e53-bdd1-6956402b2ef8"
  className="resultsComponentsItemAccordion"
>
  <div
    className="resultsComponentsItem "
  >
    <div
      className="resultsComponentsItemName"
    />
    <div
      className="resultsComponentsItemValue"
    />
    <div
      className="resultsComponentsItemRange"
    />
    <div
      className="resultsComponentsItemAdditional"
    >
      <CustomAccordionToggle
        eventKey="06cd80cf-7230-4e53-bdd1-6956402b2ef8-PAST_RESULTS"
      >
        <Component />
      </CustomAccordionToggle>
    </div>
  </div>
  <AccordionCollapse
    eventKey="06cd80cf-7230-4e53-bdd1-6956402b2ef8-NOTES"
  >
    <div
      className="resultsComponentsItemExpand"
    >
      <span
        className="resultsComponentsItemExpandTitle"
      >
        notes
      </span>
      <div
        className="resultsComponentsItemExpandText"
      />
    </div>
  </AccordionCollapse>
  <AccordionCollapse
    eventKey="06cd80cf-7230-4e53-bdd1-6956402b2ef8-PAST_RESULTS"
  >
    <div
      className="resultsComponentsItemExpand"
    >
      <div
        className="resultsComponentsItemExpandChart"
      >
        <LevelsChart
          chartData={
            Object {
              "chartOptions": Object {
                "chartScales": Object {
                  "yAxis": Object {
                    "max": 110,
                    "min": 70,
                    "stepSize": 10,
                  },
                },
              },
              "datasets": Array [
                Object {
                  "data": Object {
                    "dimension": "mg/dl",
                    "values": Array [
                      Object {
                        "x": "2021-12-15T22:44:00",
                        "y": 90.7,
                      },
                      Object {
                        "x": "2021-11-15T22:44:00",
                        "y": 95.7,
                      },
                      Object {
                        "x": "2021-10-15T22:44:00",
                        "y": 97.7,
                      },
                      Object {
                        "x": "2021-09-15T22:44:00",
                        "y": 100.7,
                      },
                      Object {
                        "x": "2021-08-15T22:44:00",
                        "y": 105.7,
                      },
                      Object {
                        "x": "2021-07-15T22:44:00",
                        "y": 95.7,
                      },
                      Object {
                        "x": "2021-06-15T22:44:00",
                        "y": 108.7,
                      },
                      Object {
                        "x": "2021-05-15T22:44:00",
                        "y": 107.7,
                      },
                      Object {
                        "x": "2021-04-15T22:44:00",
                        "y": 80.7,
                      },
                      Object {
                        "x": "2021-03-15T22:44:00",
                        "y": 85.7,
                      },
                      Object {
                        "x": "2021-02-15T22:44:00",
                        "y": 90.7,
                      },
                      Object {
                        "x": "2021-01-15T22:44:00",
                        "y": 100.7,
                      },
                      Object {
                        "x": "2022-01-15T22:44:00",
                        "y": 100.7,
                      },
                      Object {
                        "x": "2022-02-15T22:44:00",
                        "y": 100.7,
                      },
                    ],
                  },
                  "label": "Past Results",
                },
              ],
            }
          }
          chartId="past result chart"
          height={310}
        />
      </div>
    </div>
  </AccordionCollapse>
</Accordion>
`;
